File: src/backend/ai/digitalTwin.js
// backend/ai/digitalTwin.js

/**
 * Digital Twin: prediksi kondisi keuangan user 90 hari ke depan
 * Analoginya seperti GPS — bukan hanya "kamu ada di sini",
 * tapi "kalau terus begini, 3 bulan lagi kamu sampai di sini"
 */

function buildDigitalTwin(transactions, userData) {
    const { monthlyIncome, totalSavings } = userData;

    // Step 1: Hitung pengeluaran harian dari data transaksi
    const dailySpending = aggregateDailySpending(transactions);

    // Step 2: Fit linear regression (y = mx + b)
    const { slope, intercept } = linearRegression(dailySpending);

    // Step 3: Forecast 90 hari ke depan
    const forecast = forecastNext90Days(dailySpending.length, slope, intercept);

    // Step 4: Hitung proyeksi tabungan
    const savingsProjection = projectSavings(
        forecast, monthlyIncome, totalSavings
    );

    // Step 5: Deteksi warning bulan-bulan berbahaya
    const warnings = detectWarnings(savingsProjection);

    return {
        currentDailyAvg: parseFloat(calculateAverage(dailySpending).toFixed(0)),
        trend: slope > 500 ? 'NAIK' : slope < -500 ? 'TURUN' : 'STABIL',
        forecast: forecast.map(f => ({
        day: f.day,
        projectedDailySpending: Math.max(0, parseFloat(f.value.toFixed(0)))
        })),
        savingsProjection,
        warnings,
        summary: generateSummary(slope, savingsProjection, warnings)
    };
}

// ─── Helper Functions ────────────────────────────────────────

function aggregateDailySpending(transactions) {
  // Group transaksi per hari, jumlahkan
    const byDay = {};
    transactions.forEach(t => {
        const day = t.date.split('T')[0]; // ambil tanggal saja
        byDay[day] = (byDay[day] || 0) + t.amount;
    });

    // Sort by date, return sebagai array
    return Object.keys(byDay)
        .sort()
        .map(date => byDay[date]);
}

function linearRegression(data) {
    const n = data.length;
    if (n === 0) return { slope: 0, intercept: 0 };

    const xSum = data.reduce((sum, _, i) => sum + i, 0);
    const ySum = data.reduce((sum, y) => sum + y, 0);
    const xySum = data.reduce((sum, y, i) => sum + i * y, 0);
    const x2Sum = data.reduce((sum, _, i) => sum + i * i, 0);

    const denom = n * x2Sum - xSum * xSum;
    if (denom === 0) return { slope: 0, intercept: ySum / n };

    const slope = (n * xySum - xSum * ySum) / denom;
    const intercept = (ySum - slope * xSum) / n;

    return { slope, intercept };
}

function forecastNext90Days(startDay, slope, intercept) {
    const forecast = [];
    for (let i = 0; i < 90; i++) {
        forecast.push({
        day: startDay + i,
        value: slope * (startDay + i) + intercept
        });
    }
    return forecast;
}

function projectSavings(forecast, monthlyIncome, currentSavings) {
    const dailyIncome = monthlyIncome / 30;
    let savings = currentSavings;
    const projection = [];

    // Groupkan per bulan (30 hari)
    for (let month = 0; month < 3; month++) {
        const monthDays = forecast.slice(month * 30, (month + 1) * 30);
        const totalSpending = monthDays.reduce((sum, d) => sum + Math.max(0, d.value), 0);
        const totalIncome = dailyIncome * 30;
        savings += totalIncome - totalSpending;

        projection.push({
        month: month + 1,
        label: getMonthLabel(month),
        projectedSavings: Math.round(savings),
        projectedSpending: Math.round(totalSpending),
        netCashFlow: Math.round(totalIncome - totalSpending)
        });
    }

    return projection;
}

function detectWarnings(savingsProjection) {
    const warnings = [];

    savingsProjection.forEach(month => {
        if (month.projectedSavings < 0) {
        warnings.push({
            month: month.month,
            type: 'DEFISIT',
            message: `Bulan ${month.label}: tabungan bisa minus Rp ${Math.abs(month.projectedSavings).toLocaleString('id-ID')}`
        });
        } else if (month.netCashFlow < 0) {
        warnings.push({
            month: month.month,
            type: 'CASH_FLOW_NEGATIF',
            message: `Bulan ${month.label}: pengeluaran melebihi income sebesar Rp ${Math.abs(month.netCashFlow).toLocaleString('id-ID')}`
        });
        }
    });

    return warnings;
}

function generateSummary(slope, savingsProjection, warnings) {
    const lastMonth = savingsProjection[2];

    if (warnings.length === 0) {
        return `Proyeksi 3 bulan ke depan terlihat aman. Tabungan diperkirakan mencapai Rp ${lastMonth.projectedSavings.toLocaleString('id-ID')} di bulan ke-3.`;
    }

    return `Perhatian: ada ${warnings.length} bulan dengan cash flow negatif. ${warnings[0].message}.`;
    }

    function calculateAverage(arr) {
    if (arr.length === 0) return 0;
    return arr.reduce((a, b) => a + b, 0) / arr.length;
    }

    function getMonthLabel(index) {
    const now = new Date();
    now.setMonth(now.getMonth() + index + 1);
    return now.toLocaleString('id-ID', { month: 'long' });
    }

module.exports = { buildDigitalTwin };

File: src/backend/ai/digitalTwin.py
# ai/digital_twin.py
from datetime import datetime, timedelta
import random

def build_digital_twin(transactions: list, user: dict) -> dict:
    monthly_income = user["monthly_income"]
    total_savings  = user["total_savings"]

    daily_spending = _aggregate_daily(transactions)
    slope, intercept = _linear_regression(daily_spending)

    forecast = _forecast(len(daily_spending), slope, intercept, 90)
    savings_projection = _project_savings(forecast, monthly_income, total_savings)
    warnings = _detect_warnings(savings_projection)

    avg = sum(daily_spending) / len(daily_spending) if daily_spending else 0
    trend = "NAIK" if slope > 500 else "TURUN" if slope < -500 else "STABIL"

    return {
        "current_daily_avg": round(avg),
        "trend": trend,
        "savings_projection": savings_projection,
        "warnings": warnings,
        "summary": _summary(warnings, savings_projection)
    }

def _aggregate_daily(transactions):
    by_day = {}
    for t in transactions:
        day = t["date"][:10]
        by_day[day] = by_day.get(day, 0) + t["amount"]
    return [by_day[k] for k in sorted(by_day)]

def _linear_regression(data):
    n = len(data)
    if n == 0: return 0, 0
    x_sum  = sum(range(n))
    y_sum  = sum(data)
    xy_sum = sum(i * y for i, y in enumerate(data))
    x2_sum = sum(i * i for i in range(n))
    denom  = n * x2_sum - x_sum ** 2
    if denom == 0: return 0, y_sum / n
    slope     = (n * xy_sum - x_sum * y_sum) / denom
    intercept = (y_sum - slope * x_sum) / n
    return slope, intercept

def _forecast(start, slope, intercept, days):
    return [max(0, slope * (start + i) + intercept) for i in range(days)]

def _project_savings(forecast, monthly_income, current_savings):
    daily_income = monthly_income / 30
    savings = current_savings
    projection = []
    now = datetime.now()
    for month in range(3):
        month_days    = forecast[month*30:(month+1)*30]
        total_spending = sum(month_days)
        total_income   = daily_income * 30
        savings       += total_income - total_spending
        label = (now + timedelta(days=30*(month+1))).strftime("%B")
        projection.append({
            "month": month + 1,
            "label": label,
            "projected_savings": round(savings),
            "projected_spending": round(total_spending),
            "net_cash_flow": round(total_income - total_spending)
        })
    return projection

def _detect_warnings(projection):
    warnings = []
    for m in projection:
        if m["net_cash_flow"] < 0:
            warnings.append({
                "month": m["month"],
                "type": "CASH_FLOW_NEGATIF",
                "message": f"Bulan {m['label']}: pengeluaran melebihi income sebesar Rp {abs(m['net_cash_flow']):,}"
            })
    return warnings

def _summary(warnings, projection):
    if not warnings:
        last = projection[-1]
        return f"Proyeksi 3 bulan aman. Tabungan diperkirakan Rp {last['projected_savings']:,} di bulan ke-3."
    return f"Perhatian: {len(warnings)} bulan dengan cash flow negatif. {warnings[0]['message']}."

def generate_demo_transactions():
    transactions = []
    base = datetime.now() - timedelta(days=30)
    for i in range(30):
        date = (base + timedelta(days=i)).isoformat()
        amount = max(0, 50_000 + i * 1_500 + (random.random() - 0.5) * 20_000)
        transactions.append({"date": date, "amount": amount})
    return transactions